src/main/scala/temple/DSL/semantics/Validator.scala

```diff
@@ -0,0 +1,96 @@
+package temple.DSL.semantics
+
+import temple.ast.AttributeType._
+import temple.ast.{Metadata, _}
+
+private class Validator private (templefile: Templefile) {
+
+  private lazy val allStructs: Set[String] =
+    templefile.services.flatMap { case serviceName -> service => Iterator(serviceName) ++ service.structs.keys }.toSet
+
+  private lazy val allServices: Set[String] = templefile.services.keys.toSet
+
+  private def validateService(service: ServiceBlock)(context: SemanticContext): Unit = {
+    service.structs.foreachEntry(context(validateStruct))
+    service.attributes.foreachEntry(context(validateAttribute))
+    service.metadata.foreach(validateMetadata(_, service.attributes)(context))
+  }
+
+  private def validateStruct(struct: StructBlock)(context: SemanticContext): Unit = {
+    struct.attributes.foreachEntry(context(validateAttribute))
+    struct.metadata.foreach(validateMetadata(_, struct.attributes)(context))
+  }
+
+  private def validateMetadata(metadata: Metadata, attributes: Map[String, Attribute] = Map())(
+    context: SemanticContext,
+  ): Unit =
+    metadata match {
+      case _: Metadata.TargetLanguage    => // Nothing to validate yet
+      case Metadata.TargetAuth(_)        => context.fail(s"TODO: figure out how this auth works")
+      case _: Metadata.ServiceLanguage   => // Nothing to validate yet
+      case _: Metadata.Database          => // Nothing to validate yet
+      case _: Metadata.ServiceAuth       => // Nothing to validate yet
+      case _: Metadata.Readable          => // Nothing to validate yet
+      case _: Metadata.Writable          => // Nothing to validate yet
+      case Metadata.Omit(_)              => // Nothing to validate yet
+      case Metadata.ServiceEnumerable(_) => // Nothing to validate yet
+      case _: Metadata.Provider          => // Nothing to validate yet
+      case Metadata.Uses(services) =>
+        for (service <- services if !allServices.contains(service)) context.fail(s"No such service $service")
+    }
+
+  private def validateAttribute(attribute: Attribute)(context: SemanticContext): Unit = {
+    validateAttributeType(attribute.attributeType)(context)
+    attribute.accessAnnotation match {
+      case Some(Annotation.Client)    => // nothing to validate
+      case Some(Annotation.Server)    => // nothing to validate
+      case Some(Annotation.ServerSet) => // nothing to validate
+      case None                       => // nothing to validate
+    }
+  }
+
+  private def validateAttributeType(attributeType: AttributeType)(context: SemanticContext): Unit =
+    attributeType match {
+      case ForeignKey(references) if !allStructs.contains(references) =>
```

Comment:
If I read this right, this allows you to have:

```
Foo: service {...}

Bar: service {
    ...
    Baz: struct {
      foo: Foo
    }
}
```

I haven't put much though into exactly what that means, but are we ok to support that? Since at that stage they're not technically foreign keys because we can't implement that constraint over 2 databases. @smithwjv thoughts?

Reply:
And likewise `Foo.field` can reference `Baz`. Is this desirable?

Reply:
Let's leave it like that for now, and restrict it later if we need to. I don't think we've actually got an example of that in Go, so let's try and work one out in our examples and circle back on this